pkg/apiserver/handlers/featuregates/handler.go

```diff
@@ -0,0 +1,136 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package featuregates
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"gopkg.in/yaml.v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/antctl/runtime"
+	"github.com/vmware-tanzu/antrea/pkg/controller/querier"
+	"github.com/vmware-tanzu/antrea/pkg/features"
+	"github.com/vmware-tanzu/antrea/pkg/util/env"
+)
+
+var controllerOnlyGates = []string{
+	"Traceflow",
+	"AntreaPolicy",
+	"Egress",
+	"NetworkPolicyStats",
+}
+
+type (
+	Config struct {
+		// FeatureGates is a map of feature names to bools that enable or disable experimental features.
+		FeatureGates map[string]bool `yaml:"featureGates,omitempty"`
+	}
+
+	Response struct {
+		Component      string `json:"component,omitempty"`
+		FeatureName    string `json:"name,omitempty"`
+		FeatureStatus  string `json:"status,omitempty"`
+		FeatureVersion string `json:"version,omitempty"`
+	}
+)
+
+// HandleFunc returns the function which can handle queries issued by 'antctl get featuregates' command.
+// The handler function populates Antrea featuregates information to the response.
+// For now, it will return all feature gates included in features.DefaultAntreaFeatureGates for agent
+// We need to exclude any new feature gates which is not consumed by agent in the future.
+func HandleFunc(cq querier.ControllerQuerier) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		k8sClientset := cq.GetK8sClient()
+		antreaConfigName := env.GetAntreaConfigMapName()
+		antreaConfig, err := k8sClientset.CoreV1().ConfigMaps(env.GetAntreaNamespace()).Get(context.TODO(), antreaConfigName, metav1.GetOptions{})
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			klog.Errorf("Error when getting config map %s: %v", antreaConfigName, err)
+			return
+		}
+
+		agentConfig := &Config{}
+		controllerConfig := &Config{}
+		err = yaml.Unmarshal([]byte(antreaConfig.Data["antrea-agent.conf"]), agentConfig)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			klog.Errorf("Failed to unmarshal Antrea antrea-agent.conf: %v", err)
+			return
+		}
+		err = yaml.Unmarshal([]byte(antreaConfig.Data["antrea-controller.conf"]), controllerConfig)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			klog.Errorf("Failed to unmarshal Antrea antrea-controller.conf: %v", err)
+			return
+		}
+
+		agentfeatureGates := getGatesResponse(agentConfig, "agent")
+		controllerfeatureGates := getGatesResponse(controllerConfig, "controller")
```

Comment:
For controller feature gates should we still get from DefaultFeatureGate?

Reply:
@jianjuns  do you mean for controller only? since we have already got the data from configmap both for agent and controller, so I use the data from configmap to get feature gate info to make the source consistent. and I didn't see there is a need that we have to use  DefaultFeatureGate.

Reply:
Yes, I meant controller only, for 1) it is more accurate; 2) at least for controller feature gates we need not to depend on the configMap name env is set; 3) even when controller is disconnected from K8s apiserver, we can still get at least the controller features; 4) consistent with the agent implementation (I know controller can only get the agent gates from configMap though).

Reply:
@jianjuns  I didn't get it, for controller feature gates, you mean it doesn't use the data from configmap? per my understanding, it will use it which is more accurate.

Reply:
just synced with @tnqn , now I understand the case,  I will refine this part.

Reply:
@jianjuns I have removed the part to read antrea-controller.conf from configmap, so the code is getting feature gates of controller from run-time  DefaultFeatureGate directly, please help to check, thanks.

Reply:
Sorry, but I still see getGatesResponse() reads controller gates from configMap?

Reply:
yes, I kept the old getGatesResponse() and pass an empty config to the method, let me refine the code to avoid confusion.

Reply:
@jianjuns  I just refined the code and split agent and controller gatesResponse to avoid confusion, please help to review again. thanks.